[C++ for OpenCL] Explained addr space conversion in nested pointers. (#271)
KhronosGroup/OpenCL-Docs: cxx4opencl/address_spaces.txt

in case it is valid, otherwise compilation will fail with a diagnostic.

[source,cpp]

> ---------

Reviewer: ```suggestion
----------
```

Reviewer: Above you used 10 `-`.

 * Literal expressions;
 * Uniform initialization syntax `{}`;
 * Using implicit constructors.

==== Nested pointers

> {cpp} for OpenCL doesn't allow implicit address space conversions in nested

Reviewer 2: ```suggestion
{cpp} for OpenCL does not allow implicit address space conversions in nested
```

> pointers even with compatible address spaces. The following rules will apply

Reviewer 2: ```suggestion
pointers even with compatible address spaces. The following rules apply
```

> when converting between address spaces in nested pointers:

Reviewer 3: Should the rationale be documented?

Author: I don't think we should go into this. We haven't done it for any other feature and it's better to keep it focused on just documenting the language rather than covering reasonings/etc.

I can see such things can be covered in forums or other discussions, or documents.

We have discussions in OS tooling bugs regarding this.

Reviewer 3: Yes, maybe this is better done in a separate document if we don't think we can fit it here. I'm afraid that these reasonings will get lost quickly if there's no central document to pull all of them together.

Author: Right now they are all in Clang reviews, but we can certainly thinkof another document where we can explain more motivation and reasoning.

 * Implicit conversions of address spaces in nested pointers are disallowed.
 * Any address space conversion in nested pointers with safe casts (e.g.
   `const_cast`, `static_cast`, `addrspace_cast`) is disallowed.
 * Any address space conversion in nested pointers can be done using low level
   C-style or `reinterpret_cast`. No compatibility check is performed for
   address spaces in nested pointers.

[source,cpp]
----------
local int * * locgenptr;
constant int * * cnstgenptr;
int * * gengenptr;
gengenptr = const_cast<int**>(locgenptr); // illegal.
gengenptr = static_cast<int**>(cnstgenptr); // illegal.
gengenptr = reinterpret_cast<int**>(locgenptr); // legal.
gengenptr = reinterpret_cast<int**>(cnstgenptr); // legal.

> ----------

Reviewer 3: Is it worth adding an example with `addrspace_cast` as well?

Author: Sure.
https://github.com/KhronosGroup/OpenCL-Docs/pull/271/commits/5c7f296f3e25ed78a8dd5331f99ba4d854c3309f

Reviewer 3: Thanks!

 * Literal expressions;
 * Uniform initialization syntax `{}`;
 * Using implicit constructors.

==== Nested pointers

> C++ for OpenCL doesn't allow implicit address space conversions in nested

Reviewer: ```suggestion
{cpp} for OpenCL doesn't allow implicit address space conversions in nested
```

pointers even with compatible address spaces. The following rules will apply
when converting between address spaces in nested pointers:

 * Implicit conversions of address spaces in nested pointers are disallowed.
 * Any address space conversion in nested pointers with safe casts (e.g.
   `const_cast`, `static_cast`, `addrspace_cast`) is disallowed.
 * Any address space conversion in nested pointers can be done using low level
   C-style or `reinterpret_cast`. No compatibility check is performed for
   address spaces in nested pointers.

[source,cpp]

> ---------

Reviewer: ```suggestion
----------
```

local int * * locgenptr;
constant int * * cnstgenptr;
int * * gengenptr;
gengenptr = const_cast<int**>(locgenptr); // illegal.
gengenptr = static_cast<int**>(cnstgenptr); // illegal.
gengenptr = reinterpret_cast<int**>(locgenptr); // legal.
gengenptr = reinterpret_cast<int**>(cnstgenptr); // legal.

> ---------

Reviewer: ```suggestion
----------
```
